### PR TITLE
[Communication] - All Communication SDKs - Consolidate Recorded Test Sanitizers

### PR DESCRIPTION
**Pull request by minnieliu**
There was a lot of duplicated code between the recorded test sanitizers for administration, chat, and sms SDKs. 
This PR creates a centralized sanitizer and cleans up the duplicate sanitization logic.

### CODE REVIEW

File: sdk/communication/Azure.Communication.Administration/tests/Azure.Communication.Administration.Tests.csproj [hunk 1]
```diff
@@ -18,6 +18,8 @@
   <!-- Shared source from Azure.Core -->
   <ItemGroup>
     <Compile Include="$(AzureCoreSharedSources)ConnectionString.cs" Link="Shared\%(RecursiveDir)\%(Filename)%(Extension)" />
+    <Compile Include="..\..\Shared\src\CommunicationRecordedTestSanitizer.cs" Link="Shared\Communication\CommunicationRecordedTestSanitizer.cs" />
```
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Administration/tests/Azure.Communication.Administration.Tests.csproj:21]: We could put this in the `Azure.Communication.Common.Tests` project and then add reference. Having shared internal types are mainly important for public libraries.

File: sdk/communication/Azure.Communication.Administration/tests/Azure.Communication.Administration.Tests.csproj [hunk 2]
```diff
@@ -18,6 +18,8 @@
...
+
```
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Administration/tests/Azure.Communication.Administration.Tests.csproj]: nit: extra new line

File: sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs [hunk 3]
```diff
@@ -26,6 +34,11 @@ public override void SanitizeHeaders(IDictionary<string, string[]> headers)
         {
             if (headers.ContainsKey(HttpHeader.Names.Authorization))
             {
+                if (headers.ContainsKey(HttpHeader.Names.UserAgent) && headers[HttpHeader.Names.UserAgent].Any(x => x.Contains("Communication.Chat")))
```
Comment (RezaJooyandeh) [inline sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs:35]: I think it still makes sense to separate the individual libraries constants and custom logic like this.

For example we could have this class as the base type and in each library if we need more customization we can override some of the functions and add the customization.

File: sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs [hunk 4]
```diff
@@ -1,19 +1,27 @@
...
-using System.Collections.Generic;
-using System.Text.RegularExpressions;
 using Azure.Core;
 using Azure.Core.TestFramework;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
 
-namespace Azure.Communication.Sms.Tests
+namespace Azure.Communication.Pipeline
```
Comment (RezaJooyandeh) [inline sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs:10]: Interesting choice of namespace. I like it!

File: sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs [hunk 5]
```diff
@@ -1,19 +1,27 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-using System.Collections.Generic;
-using System.Text.RegularExpressions;
 using Azure.Core;
 using Azure.Core.TestFramework;
+using System.Collections.Generic;
```
Comment (RezaJooyandeh) [inline sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs:6]: nit: namespace soring is not correct. System ones should come first.

File: sdk/communication/Azure.Communication.Chat/tests/Infrastructure/ChatLiveTestBase.cs [hunk 6]
```diff
@@ -14,7 +15,7 @@ public class ChatLiveTestBase : RecordedTestBase<ChatTestEnvironment>
     {
         public ChatLiveTestBase(bool isAsync) : base(isAsync)
         {
-            Sanitizer = new ChatRecordedTestSanitizer();
+            Sanitizer = new CommunicationRecordedTestSanitizer();
```
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Chat/tests/Infrastructure/ChatLiveTestBase.cs:18]: nit: We can convert this to expression body constructor to make it look more like other parts of our code

File: sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs [hunk 7]
```diff
@@ -1,19 +1,27 @@
...
     {
         private static readonly Regex s_azureResourceRegEx = new Regex(@"[^/]+?(?=(.communication.azure))", RegexOptions.Compiled);
         private static readonly Regex s_identityInRouteRegEx = new Regex(@"(?<=identities/)([^/]+)", RegexOptions.Compiled);
+        internal const string ConnectionStringEnvironmentVariableName = "COMMUNICATION_CONNECTION_STRING";
+        internal const string ToPhoneNumberEnvironmentVariableName = "TO_PHONE_NUMBER";
+        internal const string FromPhoneNumberEnvironmentVariableName = "FROM_PHONE_NUMBER";
+        /// <summary>
+        /// This is a testing/unsigned token required on the sanitized payloads for the playback mode due to format validation on CommunicationUserCredential constructors.
+        /// </summary>
+        internal const string SanitizedChatAuthHeaderValue = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c";
```
Comment (RezaJooyandeh) [inline sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs:20]: Let's remove the comment and encapsulate it in the name: `SanitizedUnsignedUserTokenValue`

File: sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs [hunk 8]
```diff
@@ -38,9 +51,9 @@ public override string SanitizeVariable(string variableName, string environmentV
         {
             return variableName switch
```
Comment (RezaJooyandeh) [inline sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs:50]: nit: let's convert this method to expression body as well

**Review by RezaJooyandeh — COMMENTED** [review]
Thanks for this cleanup 😊. I left a couple of comments

File: sdk/communication/Azure.Communication.Administration/tests/Azure.Communication.Administration.Tests.csproj (hunk 1 above)
Comment (minnieliu) [reply sdk/communication/Azure.Communication.Administration/tests/Azure.Communication.Administration.Tests.csproj]: Moved to common.tests. Thanks for the clarification!

File: sdk/communication/Azure.Communication.Administration/tests/Azure.Communication.Administration.Tests.csproj (hunk 2 above)
Comment (minnieliu) [reply sdk/communication/Azure.Communication.Administration/tests/Azure.Communication.Administration.Tests.csproj]: Removed. thanks!

File: sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs (hunk 3 above)
Comment (minnieliu) [reply sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs]: Done, put custom logic in the libraries' own sanitizers and only kept the shared logic here.

File: sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs (hunk 7 above)
Comment (minnieliu) [reply sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs]: Done, thanks!

File: sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs (hunk 8 above)
Comment (minnieliu) [reply sdk/communication/Shared/src/CommunicationRecordedTestSanitizer.cs]: done!

File: sdk/communication/Azure.Communication.Chat/tests/Infrastructure/ChatLiveTestBase.cs (hunk 6 above)
Comment (minnieliu) [reply sdk/communication/Azure.Communication.Chat/tests/Infrastructure/ChatLiveTestBase.cs]: done

**Review by ankitarorabit — COMMENTED** [review]
This makes sense overall, the fact that you are establishing a CommunicationRecordedTestSanitizer makes sense but I don't get why you are not inheriting everything from it .. for instance chat directly inherits from RecordedTestSanitizer which kind of undermines the importance of this abstraction.. if the current common sanitizer has more methods then what chat needs then ideally we should prune it down to the absolute common ones and override the additional stuff in the individual sanitizers.. something like this..

ChatSanitizer : CommunicationSanitizer
{
       override SanitizerHeaders()
       {
              base.SanitizeHeaders();
              // Additional Chat related sanitization

**Review by ankitarorabit — APPROVED** [review]

**Comment by minnieliu** [comment]
> This makes sense overall, the fact that you are establishing a CommunicationRecordedTestSanitizer makes sense but I don't get why you are not inheriting everything from it .. for instance chat directly inherits from RecordedTestSanitizer which kind of undermines the importance of this abstraction.. if the current common sanitizer has more methods then what chat needs then ideally we should prune it down to the absolute common ones and override the additional stuff in the individual sanitizers.. something like this..
> 
> ChatSanitizer : CommunicationSanitizer
> {
> override SanitizerHeaders()
> {
> base.SanitizeHeaders();
> // Additional Chat related sanitization

File: sdk/communication/Azure.Communication.Chat/tests/Infrastructure/ChatRecordedTestSanitizer.cs [hunk 9]
```diff
@@ -24,23 +21,20 @@ public override void SanitizeHeaders(IDictionary<string, string[]> headers)
...
                     return;
                 }
                 headers[HttpHeader.Names.Authorization] = new[] { SanitizeValue };
             }
         }
-
         public override string SanitizeVariable(string variableName, string environmentVariableValue)
-        {
-            return variableName switch
+            => variableName switch
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.Chat/tests/Infrastructure/ChatRecordedTestSanitizer.cs:33]: I like body expressions :-)

**Review by DominikMe — APPROVED** [review]

**Pull request merged**